internal/controllers/ipaddressclaim.go

```diff
@@ -163,26 +164,77 @@ func (r *IPAddressClaimReconciler) reconcile(ctx context.Context, c *ipamv1.IPAd
 
 	log = log.WithValues("pool name", pool.GetName())
 
-	address := poolutil.AddressByName(addresses, c.Name)
+	address := poolutil.AddressByName(addresses, claim.Name)
 	if address == nil {
 		var err error
-		address, err = r.allocateAddress(ctx, c, pool, addresses)
+		address, err = r.allocateAddress(ctx, claim, pool, addresses)
 		if err != nil {
 			log.Error(err, "failed to allocate address")
 			return ctrl.Result{}, err
 		}
+
+		log.Info("Allocated IPAddress for IPAddressClaim",
+			"IPAddressClaim", fmt.Sprintf("%s/%s", claim.Namespace, claim.Name),
+			"IPAddress", fmt.Sprintf("%s/%s %s", address.Namespace, address.Name, address.Spec.Address),
+			"InClusterIPPool", pool.GetName())
+	} else {
+		if err := r.ensureOwnerRefsAndFinalizer(ctx, address, claim, pool); err != nil {
+			log.Error(err, "failed to ensure owner refs and finalizer on address")
+			return ctrl.Result{}, err
+		}
+
+		log.Info("Found existing IPAddress for IPAddressClaim",
+			"IPAddressClaim", fmt.Sprintf("%s/%s", claim.Namespace, claim.Name),
+			"IPAddress", fmt.Sprintf("%s/%s %s", address.Namespace, address.Name, address.Spec.Address),
+			"InClusterIPPool", pool.GetName())
 	}
+
 	if !address.DeletionTimestamp.IsZero() {
 		// We prevent deleting IPAddresses while their corresponding IPClaim still exists since we cannot guarantee that the IP
 		// wil remain the same when we recreate it.
 		log.Info("Address is marked for deletion, but deletion is prevented until the claim is deleted as well.", "address", address.Name)
 	}
 
-	c.Status.AddressRef = corev1.LocalObjectReference{Name: address.Name}
+	claim.Status.AddressRef = corev1.LocalObjectReference{Name: address.Name}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *IPAddressClaimReconciler) ensureOwnerRefsAndFinalizer(ctx context.Context, address *ipamv1.IPAddress, claim *ipamv1.IPAddressClaim, pool genericInClusterPool) error {
+	if err := controllerutil.SetControllerReference(claim, address, r.Scheme); err != nil {
+		if _, ok := err.(*controllerutil.AlreadyOwnedError); !ok {
+			return errors.Wrap(err, "Failed to update address's claim owner reference")
+		}
+	}
+
+	if err := controllerutil.SetOwnerReference(pool, address, r.Scheme); err != nil {
+		return errors.Wrap(err, "Failed to update address's pool owner reference")
+	}
+
+	var poolRefIdx int
+	poolGVK := pool.GetObjectKind().GroupVersionKind()
+	for i, ownerRef := range address.GetOwnerReferences() {
+		if ownerRef.APIVersion == poolGVK.GroupVersion().String() &&
+			ownerRef.Kind == poolGVK.Kind &&
+			ownerRef.Name == pool.GetName() {
+			poolRefIdx = i
+		}
+	}
+
+	address.OwnerReferences[poolRefIdx].Controller = pointer.Bool(false)
+	address.OwnerReferences[poolRefIdx].BlockOwnerDeletion = pointer.Bool(true)
```

Comment:
What is this logic for?

Reply:
The helper functions that ensure the owner ref do not expose setting the Controller nor the BlockOwnerDeletion flag. The logic that creates the owner refs when the IPAddress is created sets these flags. We add those flags for consistency.
https://github.com/telekom/cluster-api-ipam-provider-in-cluster/blob/f67d96e92729d54fa51626d92d8adbb2d345527a/pkg/ipamutil/address.go#L27-L28

Reply:
Got the answer for this one.